packages/flutter_tools/lib/src/fuchsia/fuchsia_device.dart

```diff
@@ -131,20 +147,160 @@ class FuchsiaDevice extends Device {
   @override
   Future<String> get sdkNameAndVersion async => 'Fuchsia';
 
-  _FuchsiaLogReader _logReader;
   @override
   DeviceLogReader getLogReader({ApplicationPackage app}) {
-    _logReader ??= _FuchsiaLogReader(this);
-    return _logReader;
+    return _logReader ??= _FuchsiaLogReader(this);
   }
+  _FuchsiaLogReader _logReader;
 
   @override
-  DevicePortForwarder get portForwarder => null;
+  DevicePortForwarder get portForwarder {
+    return _portForwarder ??= _FuchsiaPortForwarder(this);
+  }
+  _FuchsiaPortForwarder _portForwarder;
 
   @override
   void clearLogs() {
   }
 
   @override
   bool get supportsScreenshot => false;
+
+  /// List the ports currently running a dart observatory.
+  Future<List<int>> servicePorts() async {
+    final String lsOutput = await shell('ls /tmp/dart.services');
+    return parseFuchsiaDartPortOutput(lsOutput);
+  }
+
+  /// Run `command` on the Fuchsia device shell.
+  Future<String> shell(String command) async {
+    final RunResult result = await runAsync(<String>[
+      'ssh', '-F', fuchsiaSdk.sshConfig.absolute.path, id, command]);
+    if (result.exitCode != 0) {
```

Comment:
or result.stderr is not empty?

Reply:
This actually wont work, since the first time we'll always get a message like 'stderr: Warning: Permanently added 'fe80::250:b6ff:fe1e:f80d%eth1' (ED25519) to the list of known hosts.'. I could use a regex to filter that out but i don't feel great about that